pommerman/forward_model.py

```diff
@@ -156,128 +156,256 @@ def step(actions,
-                            next_positions[agent.agent_id] = next_position
+        for bomb_num, bomb in enumerate(curr_bombs):
+            curr_board[bomb.position] = constants.Item.Passage.value
+            if bomb.is_moving():
+                desired_position = utility.get_next_position(
+                    bomb.position, bomb.moving_direction)
+                if utility.position_on_board(
+                        curr_board, desired_position) and not (
+                            curr_board[desired_position] in items_and_walls):
+                    desired_bomb_positions[bomb_num] = desired_position
+
+        # Deal with position switches:
+        # Agent <-> Agent => revert to previous position
+        # Bomb <-> Bomb => revert to previous position
+        # Agent <-> Bomb => revert Bomb only to previous position (for nice kicks)
+
+        crossings = dict()
+
+        def crossing(current, desired):
+            current_x, current_y = current
+            desired_x, desired_y = desired
+            if current_x != desired_x:
+                assert current_y == desired_y
+                return ('X', min(current_x, desired_x), current_y)
+            assert current_x == desired_x
+            return ('Y', current_x, min(current_y, desired_y))
+
+        for agent_num, agent in enumerate(alive_agents):
+            if desired_agent_positions[agent_num] != agent.position:
+                border = crossing(agent.position,
+                                  desired_agent_positions[agent_num])
+                if border in crossings:
+                    desired_agent_positions[agent_num] = agent.position
+                    num, isAgent = crossings[border]
+                    assert isAgent
+                    desired_agent_positions[num] = alive_agents[num].position
                 else:
-                    # The agent made an invalid direction.
-                    agent.stop()
-            else:
-                next_positions[agent.agent_id] = None
+                    crossings[border] = (agent_num, True)
+
+        for bomb_num, bomb in enumerate(curr_bombs):
+            if desired_bomb_positions[bomb_num] != bomb.position:
+                border = crossing(bomb.position,
+                                  desired_bomb_positions[bomb_num])
+                if border in crossings:
+                    desired_bomb_positions[bomb_num] = bomb.position
+                    num, isAgent = crossings[border]
+                    if not isAgent:
+                        desired_bomb_positions[num] = curr_bombs[num].position
+                else:
+                    crossings[border] = (bomb_num, False)
+
+        # Deal with multiple agents or multible bomb collisions on desired next position
+        # by resetting desired position to current position for everyone involved in the collision
+        agent_occupancy = np.zeros_like(curr_board)
+        bomb_occupancy = np.zeros_like(curr_board)
```

Comment:
why not just make these dicts(position: count)

Reply:
I like NumPy and honestly not sure about the dict benefits